Embargo design docs (#619)
dandi/dandi-archive: doc/design/embargo-mvp.md

# Embargo Minimum Viable Product
Due to time constraints, we are implementing the core functionality for embargoing dandisets ASAP.
This document is dedicated to the minimum viable product.
A subsequent document will describe the full suite of features.

# User experience

## Web
When creating a new dandiset, users will have checkbox to choose whether or not the dandiset should be embargoed.
If so, the user must also specify an NIH award number (https://era.nih.gov/files/Deciphering_NIH_Application.pdf).

> The award number is an optional argument that is passed to the create dandiset endpoint.

Reviewer: It's optional for the endpoint--if it's not passed, the endpoint will create a non-embargoed dandiset.

For the MVP, we're focusing on NIH awardees. I assume later we will have other institutions as well, but to get this going--just NIH.

The award number will be added as a new `Funder` on an automatically generated Contributor Organization in the metadata.

Instead of the normal `Publish` button and version browser, the DLP will instead show a `Release` button.

> Clicking it will open a confirmation modal informing the user that they are unembargoing, all their data will be publicized, there is no undoing, and that it can take some time for large dandisets.

Reviewer: We probably need to revisit all our terminology comprehensively at some point. I agree that "release" is a bit too general purpose. Perhaps we can use the word "unembargo" here--it would be quite clear in context.

> Clicking it will open a confirmation modal informing the user that they are unembargoing, all their data will be publicized, there is no undoing, and that it can take some time for large dandisets.

Author: I renamed "Release" to "Unembargo"

Confirming will lock the dandiset for the duration of the release.
Once the release finishes, the dandiset will be like any other unpublished draft dandiset.

Once created, an embargoed dandiset is only visible or searchable to owners.

> Instead of the `draft` chip, an `embargoed` chip should be used instead.

Reviewer 2: when you say chip here, i'm assuming you mean ui. will the api endpoint version parameter say `embargo` instead of `draft` or will it remain as `draft` with simply the metadata containing access info. i think i would prefer to continue using "draft" as a version identifier for this.

Reviewer 3: We could just have an additional `embargo` chip next to the draft one.

## CLI
The CLI experience will be basically unchanged.
Uploads to an embargoed dandiset will function exactly the same from an API perspective, the data simply goes to a different location.

### TODO: how to browse/download embargoed zarr files?
Credentials need to be generated, distributed, and used by the client.

# Data storage

> Embargoed assets will be stored in a separate S3 bucket.

Reviewer 2: separate bucket + separate prefix within bucket for each dandiset?

Reviewer: Daniel added text below (line 36) to cover this.

This bucket is private and not browseable by the general public.

> The API server will use the embargo bucket in exactly the same way it uses the public bucket, but it will only store embargoed data.

Author: Technically it's different because of the dandiarchive prefixes

> When releasing an embargoed dandiset, all asset data for that dandiset is copied to the public bucket.

Reviewer: The idea is that garbage collection will vacuum it up on its next pass.

When uploading a new asset to an embargoed dandiset, the server will first check if that blob has already been uploaded publically.

> If so, the public blob will be used instead of uploading the data again to the embargo bucket.

Reviewer 2: in case anyone who has questions about privacy based on hash checking, i would say that should not be a concern. the data in dandi are way too specific for bitwise things to match or for someone to try generating data that matches a hash.

Reviewer: Sounds good. Resolving (unresolve if there's more to discuss).

# Django changes

## Models
The `Dandiset` model will have an `embargo_status` field that is one of `PRIVATE`, `RELEASING`, or `PUBLIC`.
* `PUBLIC` means that the Dandiset is publically accessible and publishable.
  This is the state all Dandisets currently have.

> * `PRIVATE` means that the Dandiset is embargoed.

Reviewer 2: can we semantically match this to the access restrictions, `open`, `embargoed`, `restricted`. the reason i bring this up is that we will eventually have restricted datasets as well, which will be accessible to only those that have access, but are "public".

also we may need to have sub-qualifiers for round 2 of implementation. would that be just extended states, for example: `embargoed_openmetadata`, `restricted_openmetadata`, etc. ?

Reviewer 2: as long as embargoed means `authorization required` then the restricted explanation makes sense. so the alternative would be to change:
`completely locked down` --> `authorization required to access`
`embargoed asset data` --> `authorization required to access asset data`

Author: I renamed the fields accordingly.

>   It is not searchable or viewable to non-owners.

Reviewer: Searchable and visible to owners, [yes](https://en.wikipedia.org/wiki/Exception_that_proves_the_rule#Proving_the_existence_of_the_rule).

>   It is not searchable or viewable to non-owners.

Reviewer 4: might be worth explicit wording?

>   It is not searchable or viewable to non-owners.

Author: I added some some explicit wording

  It is not publishable.
  Manifest YAML/JSON files will be written to the embargo bucket rather than the public bucket.

> * `RELEASING` means that the Dandiset is currently transitioning from embargoed to public.

Reviewer: Or `UNEMBARGOING`.

  All modification operations will return a 400 error while the release is in progress.
  This includes metadata changes and uploads.

> A new `EmbargoedAssetBlob` model will be added.

Author: We might not need to do this

Reviewer: After discussing with Daniel, it seems like having a separate emabargoed asset blob model is worth it.

This behaves the same as a normal `AssetBlob`, but points to the embargo bucket rather than the public bucket.
It also contains a reference to the `Dandiset` it belongs to.

`Asset`s will point to exactly one `AssetBlob` or `EmbargoedAssetBlob`.
An `Asset` will be considered "embargoed" if it has an `EmbargoedAssetBlob`.
An embargoed `Asset` can only belong to the dandiset named in the `EmbargoedAssetBlob`.
The automatically populated `Asset` metadata will also set the `access` field to the appropriate value depending on whether or not the asset is embargoed.

The `Upload` model will have a new optional field `embargoed_dandiset` that points to an embargoed dandiset.

> If specified, the uploaded data will be sent to the embargoed bucket instead of the public bucket, and the `Upload` will create an `EmbargoedAssetBlob` when finalized.

Reviewer 2: i've thought about one generic change related to this. that upload should require the dandiset id to be sent as a parameter when initiating. this would let allow us to use that information and would not allow an upload without creating a dandiset (since the token authorization must give the uploader ownership on the dandiset).

Reviewer 2: this would require a breaking change to the CLI, but i think this is an easy change on both the api and cli side.

Reviewer 2: didn't quite see this around ~115. my suggestion was simply that this would be a generic endpoint change to require dandiset as a parameter to the upload initialize endpoint.

Reviewer: The relevant text can be seen [here](https://github.com/dandi/dandi-api/blob/ad8ac167cd27004de0119b54f9833e607cebf52c/doc/design/embargo-mvp.md?plain=1#L123-L133). This proposal keeps the parameter optional; if we want to make it required that would be out of scope for this MVP effort. I have filed #628 to track this.

Reviewer: (Resolving this thread; please unresolve if you want to continue the discussion, or feel free to discuss further at #628.)

## API
* Create dandiset endpoint

>   Add a new optional `award_number` field.

Reviewer 2: we have to be careful with this, since a dandiset can potentially be the result of multiple awards. thus doing it with the metadata is better. i'm not sure moving from a one to one relationship to a one to many relationship is going to be easy. so the alternate thought here is adding an `award` model and then connecting the dandiset to one  or more awards.

Author: This is solely so that we can guarantee that an award is specified when creating a new embargoed dandiset. It will be recorded in the metadata. If you had multiple awards, you would pick one to specify initially, then modify the metadata later to include the rest. That might be a subpar UX, but this is the MVP, so I think it's fine for now, and easy to swap out later.

If it's really important to have multiple awards, we could eventually add the ability to specify multiple when creating the dandiset, or we could remove that requirement completely and make users set all of their awards through the meditor.

Reviewer 2: if we are simply using the metadata model instead of adding a field, the multiple award thing works already. the only addition we would add would be in the validator that the embargoed dandiset must have an award associated with it. (ideally we would even check that the nih api returns a valid signal for the award). and the ui could also be updated to simply interface with the nih api. (i would say this isn't essential for MVP). but whether a database model field is created or simply the metadata is used should be decided and my preference would be for the metadata.

Reviewer: We agree largely with what you wrote, @satra.

To be explicit: we will design the UI to have a checkbox for creating an embargoed dataset; that will in turn make a required form field for an NIH award number appear. The client will craft a metadata blob that includes the award, and send all this information to the creation endpoint. The creation endpoint will do all the nitty gritty stuff it needs to do thereafter (as described in this document).

If it is imperative to support multiple NIH awards from the jump, it should not be too difficult to design the UI to require one award number, and show a plus button to add more. Then, the "craft a metadata blob" idea from above will simply iterate through a list of award numbers. Otherwise, we can expand to this capability when we implement the full design.

(In the full design, we will also consider the idea of validating an embargoed dandiset--the idea is that an owner could put a valid NIH award number in to get past the initial hurdle, then remove it later in order to have an indefinitely embargoed secret dandiset; by "validating" for qualities such as "does this dandiset have a valid award number", etc., we could simply "freeze" any invalid dandisets, thus preventing use until they become valid. This shouldn't be necessary for the MVP.)

Reviewer 2: just to clarify i don't think it's necessary for the create dandiset ui to have support for multiple awards, just one is sufficient there, and may be a note that additional awards can be added through the metadata editor after the dandiset is created. my comment was more specific to adding a db field that supports only one award.

Reviewer: I think there was some confusion in the original language: there was never an intent to track a single award number in the DB, separately from the metadata (but rather, a serializer field in the endpoint for dandiset creation). I think we're all on the same page now.

Author: I updated the design of the create dandiset endpoint, it now takes the award information via the metadata as requested.
Resolving

  If set, the `access` metadata field will be set to `EmbargoedAccess` and the award number will be recorded as a `Funder` on a new Contributor Organization.

>   The `embargo_status` of the dandiset will also be set to `PRIVATE`.

Reviewer: I don't think this is a good idea. Naming this field `access` will cause almost certainly cause confusion for us later, since e.g. `UNEMBARGOING`/`MAKING_PUBLIC` is *not* a metadata access level.

* Get/List dandiset endpoint

  The DandisetViewSet queryset will filter out dandisets with `embargo_status == PRIVATE` that are also not owned by the current user.
  This will prevent them from showing up in the listing or fetching endpoints.

  The `DandisetSerializer` will also be updated to include the `embargo_status` field so that the web client can render the embargoed dandiset appropriately.

* publish version endpoint

>   Return error 400 if `dandiset.embargo_status != PUBLIC`.

Reviewer: Same as above, 400 is fine, or perhaps 403.

* create asset, update metadata, and any other dandiset/version modification endpoints:

>   Return error 400 if `dandiset.embargo_status == RELEASING`.

Reviewer: 503 is almost certainly the wrong thing--that indicates that the whole server is (temporarily) borked. I don't think we need to send a Retry-After here; we can instead send a reason in the response body and allow the client to try again after a while.

>   Return error 400 if `dandiset.embargo_status == RELEASING`.

Reviewer 4: indeed -- 503 is no good, thank you @waxlamp for clarifying.  
I still think that some explicit, server provided "Retry-After" is better than some "a while" (a second? minute? ...) which might cause some eager clients retry way too often.  But it seems there is no other suitable response code with Retry-After (https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Retry-After). And since we will not be monitoring individual connections etc, we can't use [429 "Too many requests"](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/429) really

* New endpoint: `POST /api/dandisets/{dandiset_id}/release`

  Release an embargoed dandiset.
  

>   Only permitted for owners and admins. If the `embargo_status` is `PUBLIC` or `RELEASING`, return 400.

Reviewer: Technically, 405 refers to the *HTTP* method not being allowed, so let's keep this at 400 (it is a client error to request this operation without the proper authorization). 403 might be appropriate here as well.

  Set the `embargo_status` to `RELEASING`, then dispatch the release task.

* Release task

  For every `Asset` with an `EmbargoedAssetBlob` in the dandiset, convert the `EmbargoedAssetBlob` into an `AssetBlob` by moving the data from the embargo bucket to the public bucket.

>   These could be >5GB, so the [multipart copy API](https://docs.aws.amazon.com/AmazonS3/latest/userguide/CopyingObjectsMPUapi.html) must be used.

Reviewer 2: https://aws.amazon.com/premiumsupport/knowledge-center/s3-large-transfer-between-buckets/

the main thing is to ensure we use something that doesn't change the etag.

* Get/List asset endpoint

  The AssetViewSet queryset will filter out assets with `embargoed_asset_blob.dandiset.embargo_status != PUBLIC` that are also not owned by the current user.
  This will prevent them from showing up in the listing or fetching endpoints.

* upload_initialize_view

  An optional field `embargoed_dandiset` will be available on the serializer.
  If specified, it will be passed to the `Upload` object.
  This will mean the final upload will result in an `EmbargoedAssetBlob`.

  Even if `embargoed_dandiset` is specified, if an `AssetBlob` with a matching checksum already exists, return it instead of uploading embargoed data.
  This means that an embargoed dandiset can contain both embargoed and unembargoed assets.

  `EmbargoedAssetBlob`s should also be checked for deduplication, but only within an embargoed dandiset.
  An embargoed dandiset should use the same `EmbargoedAssetBlob` if the same file appears in multiple places, but two embargoed dandisets should upload the same data twice if they both contain the same file.

  Return error 400 if `dandiset.embargo_status == RELEASING`.

* TODO zarr upload

* stats_view
  

>   The total size value should include the size of `EmbargoedAssetBlob`s as well as `AssetBlob`s.

Reviewer: For the MVP, let's not worry about this. I think in the full version, it would be nice to see something like "300 TB published (of which 100 TB currently under embargo)" with `embargo` linked to an explanation.

# Embargo Minimum Viable Product
Due to time constraints, we are implementing the core functionality for embargoing dandisets ASAP.
This document is dedicated to the minimum viable product.
A subsequent document will describe the full suite of features.

# User experience

## Web
When creating a new dandiset, users will have checkbox to choose whether or not the dandiset should be embargoed.
If so, the user must also specify an NIH award number (https://era.nih.gov/files/Deciphering_NIH_Application.pdf).

> The award number is an optional argument that is passed to the create dandiset endpoint.

Reviewer: The award number is optional in terms of the endpoint, but required if a user wishes to create an embargoed dataset. The optionality is tied directly to whether or not the endpoint creates an embargoed or open dandiset, right?

That is, you never have to supply an award number, but you get an embargoed dandiset if and only if you supply one. Is that correct?

Reviewer 2: sounds about right.

> The award number is an optional argument that is passed to the create dandiset endpoint.

Reviewer: Suggest cutting this sentence to avoid implementation confusion later.

```suggestion
```

The award number will be added as a new `Funder` on an automatically generated Contributor Organization in the metadata.

Instead of the normal `Publish` button and version browser, the DLP will instead show a `Release` button.
Clicking it will open a confirmation modal informing the user that they are unembargoing, all their data will be publicized, there is no undoing, and that it can take some time for large dandisets.
Confirming will lock the dandiset for the duration of the release.

> Once the release finishes, the dandiset will be like any other unpublished draft dandiset.

Reviewer: Need to make sure the release is all or nothing, which might be a little bit tricky since it will generally involve moving a lot of data from bucket to bucket.

One approach may be to perform the full copy, validate it, and only then delete the embargoed bucket assets and finish "releasing" the dandiset.

Reviewer: We should maybe also email the owner/admins when the release is done?

Reviewer 2: since we would essentially copy from blob store to blob store, there are several s3 ways of doing that. the key is to preserve the s3 generated etag, which should also be possible. this would mean validation is simply checking the etag post copy.

Reviewer: Added some text to other parts of the doc to cover all this, resolving.

Once created, an embargoed dandiset is only visible or searchable to owners.
Instead of the `draft` chip, an `embargoed` chip should be used on the dandiset listing pages.
These dandisets+chips will only appear in the listings to owners of the embargoed dandisets.

## CLI
The CLI experience will be basically unchanged.
Uploads to an embargoed dandiset will function exactly the same from an API perspective, the data simply goes to a different location.

### TODO: how to browse/download embargoed zarr files?

> Credentials need to be generated, distributed, and used by the client.

Reviewer: Is this going to be part of the MVP?

Reviewer 2: i think credentials are part of mvp almost independent of zarr, but essential for zarr. since people should be able to access the endpoint without always going through pre-signed urls right? for example to load an hdf5 file via streaming one would need the s3 path. a presigned url doesn't work for s3 read access for all tools.

# Data storage
Embargoed assets will be stored in a separate S3 bucket.
This bucket is private and not browseable by the general public.

> Everything in the embargoed bucket will be prefixed with a dandiset identifier.

Reviewer: Trying to resolve some possible ambiguity in this sentence:

```suggestion
Each dandiset stored in the embargoed bucket will be prefixed with a dandiset identifier.
```

  This will prevent them from showing up in the listing or fetching endpoints.

  The `DandisetSerializer` will also be updated to include the `embargo_status` field so that the web client can render the embargoed dandiset appropriately.

* publish version endpoint

  Return error 400 if `dandiset.embargo_status != PUBLIC`.

* create asset, update metadata, and any other dandiset/version modification endpoints:

  Return error 400 if `dandiset.embargo_status == RELEASING`.

* New endpoint: `POST /api/dandisets/{dandiset_id}/release`

  Release an embargoed dandiset.
  
  Only permitted for owners and admins. If the `embargo_status` is `PUBLIC` or `RELEASING`, return 400.

  Set the `embargo_status` to `RELEASING`, then dispatch the release task.

* Release task

  For every `Asset` with an `EmbargoedAssetBlob` in the dandiset, convert the `EmbargoedAssetBlob` into an `AssetBlob` by moving the data from the embargo bucket to the public bucket.
  These could be >5GB, so the [multipart copy API](https://docs.aws.amazon.com/AmazonS3/latest/userguide/CopyingObjectsMPUapi.html) must be used.

>   The ETag and checksum must remain undisturbed; the only change should be where the data is stored.

Reviewer: ```suggestion
  The ETag and checksum must remain undisturbed; the only change should be where the data is stored.
  Verify that the resulting unembargoed assets match one-for-one (in the database) with the embargoed assets that were copied.
```

  Once finished, the `access` metadata field on the dandiset will be updated to `OpenAccess` and `embargo_status` is set to `PUBLIC`.
  
  Before copying data, check if an existing `AssetBlob` with the same checksum has been uploaded already (this would have happened after uploading the embargoed data).
  If so, use it instead of copying the `EmbargoedAssetBlob` data.

* Get/List asset endpoint

  The AssetViewSet queryset will filter out assets with `embargoed_asset_blob.dandiset.embargo_status != PUBLIC` that are also not owned by the current user.
  This will prevent them from showing up in the listing or fetching endpoints.

* upload_initialize_view

  An optional field `embargoed_dandiset` will be available on the serializer.
  If specified, it will be passed to the `Upload` object.
  This will mean the final upload will result in an `EmbargoedAssetBlob`.

  Even if `embargoed_dandiset` is specified, if an `AssetBlob` with a matching checksum already exists, return it instead of uploading embargoed data.
  This means that an embargoed dandiset can contain both embargoed and unembargoed assets.

>   `EmbargoedAssetBlob`s should also be checked for deduplication, but only within an embargoed dandiset.

Reviewer: ```suggestion
  `EmbargoedAssetBlob`s should also be checked for deduplication, but only within an embargoed dandiset.
  This is to keep the permission model clean for owners of different embargoed dandisets that might contain the same asset.
```

Permissions will also be checked just as when doing an asset download.

A test implementation can be found [here](https://github.com/dandi/dandi-api/commit/46f637c4d29d69d34a1a78d820193402d3fc3778).

# Django changes

## Models
The `Dandiset` model will have an `embargo_status` field that is one of `EMBARGOED`, `RELEASING`, or `OPEN`.
* `OPEN` means that the Dandiset is publically accessible and publishable.
  This is the state all Dandisets currently have.
* `EMBARGOED` means that the Dandiset is embargoed.
  It is not searchable or viewable to non-owners.
  It is not publishable.
  Manifest YAML/JSON files will be written to the embargo bucket rather than the public bucket.
* `RELEASING` means that the Dandiset is currently transitioning from embargoed to public.
  All modification operations will return a 400 error while the release is in progress.
  This includes metadata changes and uploads.

A new `EmbargoedAssetBlob` model will be added.
This behaves the same as a normal `AssetBlob`, but points to the embargo bucket rather than the public bucket.
It also contains a reference to the `Dandiset` it belongs to.

> `Asset`s will point to exactly one `AssetBlob` or `EmbargoedAssetBlob`.

Reviewer 4: ```suggestion
`Asset`s will point to exactly one `AssetBlob` or `EmbargoedAssetBlob` or `EmbargoedZarrArchive` or `ZarrArchive`.
```
? (combinatorics start to creep in)

> An `Asset` will be considered "embargoed" if it has an `EmbargoedAssetBlob`.

Reviewer 4: ```suggestion
An `Asset` will be considered "embargoed" if it has an `EmbargoedAssetBlob` or `EmbargoedZarrArchive`.
```
?

An embargoed `Asset` can only belong to the dandiset named in the `EmbargoedAssetBlob`.
The automatically populated `Asset` metadata will also set the `access` field to the appropriate value depending on whether or not the asset is embargoed.

A new `EmbargoedUpload` model will be added.
This is just like the normal `Upload` model, but the uploaded data will be sent to the embargoed bucket instead of the public bucket, and the `Upload` will create an `EmbargoedAssetBlob` when finalized.
It will also contain a reference to the `Dandiset` it belongs.

> A new `EmbargoedZarrArchive` model will be added in the same vein as `EmbargoedAssetBlob`.

Reviewer 4: should we use `*ZarrFile` terms instead of `*ZarrArchive` to stay inline with below `EmbargoedZarrUploadFile` etc?

Reviewer: I think there's a meaningful distinction: the "archive" is the full asset, while the "file" refers to a single file of the zarr (as it is uploaded). @dchiquito?

Reviewer 4: eh... I always disliked referring to `.zarr/` directory as "zarr file" but it grew on me I guess.
Also as we refer to "files" as "blobs", may be then it still could be `ZarrFile` (the "archive") and `ZarrBlob` (an individual file within Zarr)?  or should we may be check how zarr calls itself and individual files? ;)

Author: >  referring to .zarr/ directory as "zarr file"
That's the opposite of what we're doing, the entire zarr directory is the "zarr archive", the files it contains are "zarr files".

Looks like zarr might refer to the component files as "chunks", but I'm not sure if that might also refer to subdirectories.
